### PR TITLE
Update CI

### PR DESCRIPTION
**Pull request by dmsimard**

### CODE REVIEW

**Comment by dmsimard** [comment]
Got a few seemingly unrelated timeout failures.

On the nginx job we have [this](https://dashboard.zuul.ansible.com/t/ansible/build/2247a2c717af45c4b991d60fc1d7116a/console#2/0/21/centos-8):
```
# [...]
Installing collected packages: pbr, autopage, stevedore, wcwidth, PrettyTable, attrs, pyperclip, cmd2, cliff, certifi, urllib3, idna, charset-normalizer, requests, pytz, sqlparse, Django, djangorestframework, django-cors-headers, django-filter, django-health-check, ruamel.yaml.clib, ruamel.yaml, dynaconf, tzlocal, whitenoise, pygments, ara

:stderr: Exception:
Traceback (most recent call last):
  File "/usr/lib/python3.6/site-packages/pip/basecommand.py", line 215, in main
    status = self.run(options, args)
  File "/usr/lib/python3.6/site-packages/pip/commands/install.py", line 365, in run
    strip_file_prefix=options.strip_file_prefix,
  File "/usr/lib/python3.6/site-packages/pip/req/req_set.py", line 789, in install
    **kwargs
  File "/usr/lib/python3.6/site-packages/pip/req/req_install.py", line 854, in install
    strip_file_prefix=strip_file_prefix
  File "/usr/lib/python3.6/site-packages/pip/req/req_install.py", line 1069, in move_wheel_files
    strip_file_prefix=strip_file_prefix,
  File "/usr/lib/python3.6/site-packages/pip/wheel.py", line 345, in move_wheel_files
    clobber(source, lib_dir, True)
  File "/usr/lib/python3.6/site-packages/pip/wheel.py", line 316, in clobber
    ensure_dir(destdir)
  File "/usr/lib/python3.6/site-packages/pip/utils/__init__.py", line 83, in ensure_dir
    os.makedirs(path)
  File "/usr/lib64/python3.6/os.py", line 220, in makedirs
    mkdir(name, mode)
PermissionError: [Errno 13] Permission denied: '/usr/local/lib/python3.6/site-packages/pbr-5.8.0.dist-info'
```

It's because pip attempts to install systemwide using an unprivileged user. We could elevate privileges but we could also specify ``extra_vars: --user`` when the user isn't privileged ?
FWIW we have an issue about supporting ``--user``: https://github.com/ansible-community/ara-collection/issues/3

**Comment by dmsimard** [comment]
Other failures spotted in the [nginx job](https://dashboard.zuul.ansible.com/t/ansible/build/cb17b6258eb948ff900774fde2a8b6b6):

Missing ``openssl`` on fedora35 but it's there for ubuntu and centos8:
```
Generate private key and self-signed certificate

cmd

openssl req -x509 -config /etc/ssl/ara.example.org/selfsigned.conf -newkey rsa:2048 -keyform PEM -out /etc/ssl/ara.example.org/cert.pem -outform PEM -days 3650 -nodes

msg

[Errno 2] No such file or directory: 'openssl'
```

For centos8 the ``pip3 install --user ara`` works (and this is [without specifying the full path to pip3](https://github.com/ansible-community/ara-collection/pull/51/commits/75e38fa6f7e2030d832981bf0aeb23d9d2b3c7ac) @Thulium-Drake): https://dashboard.zuul.ansible.com/t/ansible/build/cb17b6258eb948ff900774fde2a8b6b6/console#2/0/21/centos-8

However for Ubuntu 18.04 it returns:
```
Unable to find any of pip3 to use.  pip needs to be installed.
```

Later, the centos8 almost finishes successfully but fails with [this error](https://dashboard.zuul.ansible.com/t/ansible/build/cb17b6258eb948ff900774fde2a8b6b6/console#2/0/52/centos-8):
```
Failed to import the required Python library (passlib) on centos-8-1vcpu-vexxhost-sjc1-0002091278's Python /usr/bin/python3. Please read module documentation and install in the appropriate location. If the required library is installed, but Ansible is using the wrong Python interpreter, please consult the documentation on ansible_python_interpreter
```

The comment says the intent is to effectively select the right python interpreter so there's a bug here:
https://github.com/ansible-community/ara-collection/blob/cfde8262df84676168e1bf23ff65cdbb7fec5098/roles/ara_frontend_nginx/tasks/main.yaml#L45-L46

passlib was installed successfully earlier in the playbook:
```
0:
"/usr/bin/pip3"
1:
"install"
2:
"--user"
3:
"passlib"
]

stdout

Collecting passlib
  Downloading https://files.pythonhosted.org/packages/3b/a4/ab6b7589382ca3df236e03faa71deac88cae040af60c071a78d254a62172/passlib-1.7.4-py2.py3-none-any.whl (525kB)
Installing collected packages: passlib
Successfully installed passlib-1.7.4

```

Could it be that ansible doesn't load the user site-packages (where ``pip install --user`` goes) ?

Need to continue troubleshooting and iterating on this PR for a bit.

**Comment by Thulium-Drake** [comment]
Python's module packaging is such a mess :see_no_evil: ... In Debian they tried to fix it, but it also sometimes doesn't work as expected :)

Isn't it because passlib is installed inside the venv, but the ansible interpreter is _not_ inside the venv when running the htpasswd commands?

**Comment by dmsimard** [comment]
> Isn't it because passlib is installed inside the venv, but the ansible interpreter is _not_ inside the venv when running the htpasswd commands?

The output from zuul made it hard to follow and refer to so I've reproduced the issue and recorded it instead: https://demo.recordsansible.org/playbooks/3231.html _(the CI job playbooks used to be recorded before migrating to a new zuul, I should fix that)_

What's happening is that we're installing passlib as an unprivileged user with ``--user``: https://demo.recordsansible.org/results/690261.html.

Then we run the [htpasswd task](https://demo.recordsansible.org/results/690291.html) with the correct ``ansible_python_interpreter`` but we run it with ``become: true`` in [this block](https://demo.recordsansible.org/files/244481.html#line-25) (to install the file in a privileged directory) so root doesn't see the unprivileged user's local python library. Makes sense.

It would probably work if I would run the entire playbook with ``become: true`` because then ``pip install --user`` would install to root's home directory. I see running the server from an unprivileged user a feature, though, so I'll figure a fix.

As a side note, that passlib installation task should probably move to the nginx role since it's the only place where it is used.

**Comment by Thulium-Drake** [comment]
> > Isn't it because passlib is installed inside the venv, but the ansible interpreter is _not_ inside the venv when running the htpasswd commands?
> 
> The output from zuul made it hard to follow and refer to so I've reproduced the issue and recorded it instead: https://demo.recordsansible.org/playbooks/3231.html _(the CI job playbooks used to be recorded before migrating to a new zuul, I should fix that)_
> 
> What's happening is that we're installing passlib as an unprivileged user with `--user`: https://demo.recordsansible.org/results/690261.html.
> 
> Then we run the [htpasswd task](https://demo.recordsansible.org/results/690291.html) with the correct `ansible_python_interpreter` but we run it with `become: true` in [this block](https://demo.recordsansible.org/files/244481.html#line-25) (to install the file in a privileged directory) so root doesn't see the unprivileged user's local python library. Makes sense.
> 
Oh.. right! Yeah, that makes sense. 
> It would probably work if I would run the entire playbook with `become: true` because then `pip install --user` would install to root's home directory. I see running the server from an unprivileged user a feature, though, so I'll figure a fix.
Indeed, however, installing and configuring a webserver can't really be done without extensive modifications as a non-root user. So I think it's wise to stick to the scope of running ARA as a non-privileged user.
> 
> As a side note, that passlib installation task should probably move to the nginx role since it's the only place where it is used.
I agree, well, in that case, let's start by moving that task to the role it belongs. I'd even say, put it in the same block.  :-)

**Comment by dmsimard** [comment]
recheck

**Comment by dmsimard** [comment]
I forced pushed the wrong branch :facepalm: but it's okay, github has the ref here for a neat venv cleanup that can go in another PR: https://github.com/ansible-community/ara-collection/commit/d6156bdee661f76a199c2b5e9c5ffcb52366d892

The error is likely due to the bump from centos stream 8 to centos stream 9: https://ansible.softwarefactory-project.io/zuul/build/b172f812bc54418ebc1d435b3b05e2cd/console#2/1/8/centos-stream-9 ``No package python36 available.``

I guess there is value in keeping testing coverage for both centos 8 and centos 9, they'll also need different vars.

File: roles/ara_api/tasks/main.yaml [hunk 1]
```diff
@@ -19,6 +19,7 @@
 - name: Include OS family/distribution specific variables
   include_vars: "{{ item }}"
   with_first_found:
+    - "{{ ansible_facts['distribution'] }}_{{ ansible_facts['ansible_distribution_major_version'] }}.yaml"
```
Comment (dmsimard) [inline roles/ara_api/tasks/main.yaml:22]: Maybe I was too optimistic: ``No file was found when using with_first_found. Use the 'skip: true' option to allow this task to be skipped if no files are found``

**Comment by dmsimard** [comment]
Progress !

mysql and postgresql are fixed and the nginx failure looks unrelated (mirror dns resolution error).

I reproduced the problem with the distributed-sqlite job locally -- ansible starts the gunicorn service but it fails by tripping over itself due to the logging configuration (reported in https://github.com/ansible-community/ara-collection/issues/55):
```
Aug 01 00:18:55 centos9s-testing.novalocal gunicorn[17958]: [2022-08-01 04:18:55 +0000] [17958] [INFO] Worker exiting (pid: 17958)
Aug 01 00:18:55 centos9s-testing.novalocal gunicorn[17956]: [ara] Server's LOGGING configuration has a 'root' logger which is deprecated after version 1.5.8 due to conflicts. Please update your settings.yaml or generate a new one by removing it from ara's configuration folder
Aug 01 00:18:55 centos9s-testing.novalocal gunicorn[17958]: [ara] Server's LOGGING configuration has a 'root' logger which is deprecated after version 1.5.8 due to conflicts. Please update your settings.yaml or generate a new one by removing it from ara's configuration folder
Aug 01 00:18:55 centos9s-testing.novalocal gunicorn[17959]: [2022-08-01 04:18:55 +0000] [17959] [ERROR] Exception in worker process
Aug 01 00:18:55 centos9s-testing.novalocal gunicorn[17959]: Traceback (most recent call last):
Aug 01 00:18:55 centos9s-testing.novalocal gunicorn[17959]:   File "/usr/lib64/python3.9/logging/config.py", line 564, in configure
Aug 01 00:18:55 centos9s-testing.novalocal gunicorn[17959]:     handler = self.configure_handler(handlers[name])
Aug 01 00:18:55 centos9s-testing.novalocal gunicorn[17959]:   File "/usr/lib64/python3.9/logging/config.py", line 745, in configure_handler
Aug 01 00:18:55 centos9s-testing.novalocal gunicorn[17959]:     result = factory(**kwargs)
Aug 01 00:18:55 centos9s-testing.novalocal gunicorn[17959]:   File "/usr/lib64/python3.9/logging/handlers.py", line 211, in __init__
Aug 01 00:18:55 centos9s-testing.novalocal gunicorn[17959]:     BaseRotatingHandler.__init__(self, filename, 'a', encoding=encoding,
Aug 01 00:18:55 centos9s-testing.novalocal gunicorn[17959]:   File "/usr/lib64/python3.9/logging/handlers.py", line 58, in __init__
Aug 01 00:18:55 centos9s-testing.novalocal gunicorn[17959]:     logging.FileHandler.__init__(self, filename, mode=mode,
Aug 01 00:18:55 centos9s-testing.novalocal gunicorn[17959]:   File "/usr/lib64/python3.9/logging/__init__.py", line 1146, in __init__
Aug 01 00:18:55 centos9s-testing.novalocal gunicorn[17959]:     StreamHandler.__init__(self, self._open())
Aug 01 00:18:55 centos9s-testing.novalocal gunicorn[17959]:   File "/usr/lib64/python3.9/logging/__init__.py", line 1175, in _open
Aug 01 00:18:55 centos9s-testing.novalocal gunicorn[17959]:     return open(self.baseFilename, self.mode, encoding=self.encoding,
Aug 01 00:18:55 centos9s-testing.novalocal gunicorn[17959]: PermissionError: [Errno 13] Permission denied: '/home/centos/.ara-tests/logs/server.log'
```

I'll fix it.

**Comment by dmsimard** [comment]
Fixed the logging issue -- the permission denied was actually because of selinux being enforcing and the gunicorn policies not working correctly.

**Comment by dmsimard** [comment]
Jobs are successful now but by adding centos9 I removed centos8 -- I'll add it back in so we keep coverage for it.

**Comment by dmsimard** [comment]
recheck

File: roles/ara_api/vars/CentOS_9.yaml [hunk 2]
```diff
@@ -0,0 +1,44 @@
...
+#  the Free Software Foundation, either version 3 of the License, or
+#  (at your option) any later version.
+#
+#  ARA Records Ansible is distributed in the hope that it will be useful,
+#  but WITHOUT ANY WARRANTY; without even the implied warranty of
+#  MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+#  GNU General Public License for more details.
+#
+#  You should have received a copy of the GNU General Public License
+#  along with ARA Records Ansible. If not, see <http://www.gnu.org/licenses/>.
```
Comment (dmsimard) [inline roles/ara_api/vars/CentOS_9.yaml:17]: Unrelated: we've switched to short-form GPL license headers in ara, I'll carry them over here so the boilerplate isn't as long.

**Comment by dmsimard** [comment]
Looks good to me, this will allow us to unblock merging of existing PRs but they will need to be rebased.
